dvc/data/stage.py

```diff
@@ -247,6 +247,24 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
+def _load_untyped_dir_hash(odb, fs, fs_path, hash_info, meta, staging):
+    from .tree import Tree
+
+    obj = Tree.load(odb, hash_info.as_raw())
+    obj.hash_info = hash_info
+    meta.nfiles = len(obj)
+
+    staging.add(obj.fs_path, obj.fs, obj.hash_info)
+    for key, (_, entry_hash_info) in obj.iteritems():
+        staging.add(
+            fs.path.join(fs_path, *key),
+            fs,
+            entry_hash_info,
+            verify=False,
+        )
+    return staging, meta, Tree.load(staging, hash_info)
```

Comment:
Seems like this is doing much more than "load untyped dir hash". Also "load hash" is not really describing what is going on here. Maybe there is a way to clean the code up?

Reply:
How about `_stage_tree_from_untyped_hash`?

Reply:
Also note that you are loading the tree twice (which means that you are reading and parsing json twice). Do you see that taking significant time in the benchmarks? If so, is there a nice way to avoid that?